vpp/oclpostprocess_transform.cpp

```diff
@@ -0,0 +1,290 @@
+            status = YAMI_SUCCESS;
+        }
+    } break;
+    default:
+        status = OclPostProcessBase::setParameters(type, vppParam);
+        break;
+    }
+    return status;
+}
+
+YamiStatus OclPostProcessTransform::flip(const SharedPtr<OclVppCLImage>& src,
+    const SharedPtr<OclVppCLImage>& dst)
+{
+    uint32_t width = src->getWidth();
+    uint32_t height = src->getHeight();
+    if (width != dst->getWidth() || height != dst->getHeight()) {
+        ERROR("flip failed due to unmatched resolution");
+        return YAMI_INVALID_PARAM;
+    }
+
+    uint32_t size;
+    cl_kernel kernel = NULL;
+    if (m_transform & VPP_TRANSFORM_FLIP_H) {
+        size = width / 4 - 1;
+        kernel = getKernel("transform_flip_h");
+    }
+    else if (m_transform & VPP_TRANSFORM_FLIP_V) {
+        size = height;
+        kernel = getKernel("transform_flip_v");
+    }
+    if (!kernel) {
+        ERROR("failed to get cl kernel");
+        return YAMI_FAIL;
+    }
+    if ((CL_SUCCESS != clSetKernelArg(kernel, 0, sizeof(cl_mem), &dst->plane(0)))
+         || (CL_SUCCESS != clSetKernelArg(kernel, 1, sizeof(cl_mem), &dst->plane(1)))
+         || (CL_SUCCESS != clSetKernelArg(kernel, 2, sizeof(cl_mem), &src->plane(0)))
+         || (CL_SUCCESS != clSetKernelArg(kernel, 3, sizeof(cl_mem), &src->plane(1)))
+         || (CL_SUCCESS != clSetKernelArg(kernel, 4, sizeof(uint32_t), &size))) {
+        ERROR("clSetKernelArg failed");
+        return YAMI_FAIL;
+    }
+
+    size_t globalWorkSize[2], localWorkSize[2];
+    localWorkSize[0] = 8;
+    localWorkSize[1] = 8;
+    globalWorkSize[0] = ALIGN_POW2(width, localWorkSize[0] * 4) / 4;
+    globalWorkSize[1] = ALIGN_POW2(height, localWorkSize[1] * 2) / 2;
+    if (!checkOclStatus(clEnqueueNDRangeKernel(m_context->m_queue, kernel, 2, NULL,
+                            globalWorkSize, localWorkSize, 0, NULL, NULL),
+            "EnqueueNDRangeKernel")) {
+        return YAMI_FAIL;
+    }
+    return YAMI_SUCCESS;
+}
+
+YamiStatus OclPostProcessTransform::rotate(const SharedPtr<OclVppCLImage>& src,
+    const SharedPtr<OclVppCLImage>& dst)
+{
+    uint32_t width = src->getWidth();
+    uint32_t height = src->getHeight();
+
+    uint32_t size, w, h;
+    cl_kernel kernel = NULL;
+    if (m_transform & VPP_TRANSFORM_ROT_90) {
+        if (width != dst->getHeight() || height != dst->getWidth()) {
+            ERROR("rotate failed due to unmatched resolution");
+            return YAMI_INVALID_PARAM;
+        }
+        size = 4;
+        w = width / 4 - 1;
+        h = height / 4 - 1;
+        kernel = getKernel("transform_rot_90");
+    }
+    else if (m_transform & VPP_TRANSFORM_ROT_180) {
+        if (width != dst->getWidth() || height != dst->getHeight()) {
+            ERROR("rotate failed due to unmatched resolution");
+            return YAMI_INVALID_PARAM;
+        }
+        size = 2;
+        w = width / 4 - 1;
+        h = height;
+        kernel = getKernel("transform_rot_180");
```

Comment:
This maybe have performance issue, if you compare the string for every frame.best way is cache kernels. But you can update this later